### PR TITLE
Use month codes in formatting

### LINKED ISSUE #486: Implement month and era codes
The syntax for month and era codes is being discussed in #355.  I implemented a placeholder in #445.  This ticket is to follow up and properly wire month codes throughout the data model and document them appropriately.

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #486

This moves everyone over to using month codes for formatting. As discussed in #486, ordinal months are still used when requesting a numerical month format, though well-formed data for lunar calendars should never contain such skeletons.

The datagen macro needed to be updated a bit to work with this but I think it turned out well!

The regen commit is massive, avoid opening it in your browser and instead review commit by commit.

testdata.postcard goes up in size by 0.03 MB, due to all of the flat arrays turning into maps. This is understandable, we have a _lot_ of calendar files.

### CODE REVIEW

File: components/datetime/src/provider/calendar/symbols.rs [hunk 1]
```diff
@@ -170,22 +170,24 @@ symbols!(
...
-            Cow::Owned("June".to_string()),
-            Cow::Owned("July".to_string()),
-            Cow::Owned("August".to_string()),
-            Cow::Owned("September".to_string()),
-            Cow::Owned("October".to_string()),
-            Cow::Owned("November".to_string()),
-            Cow::Owned("December".to_string()),
-        ]);
+        let months = [
+            (&MonthCode(tinystr!(4, "M01")), "January"),
```
Comment (Manishearth) [inline components/datetime/src/provider/calendar/symbols.rs:177]: I think we should have a more convenient way of constructing month codes. How about a monthcode macro?

File: provider/datagen/src/transform/cldr/datetime/symbols.rs [hunk 2]
```diff
@@ -63,37 +92,37 @@ fn get_era_code_map(calendar: &str) -> BTreeMap<String, TinyStr16> {
         ]
         .into_iter()
         .collect(),
-        #[allow(clippy::panic)] // TODO(#1668) Clippy exceptions need docs or fixing.
+        #[allow(clippy::panic)] // Panics okay in datagen
```
Comment (robertbastian) [inline provider/datagen/src/transform/cldr/datetime/symbols.rs:95]: we don't have the `deny` in this crate

File: components/datetime/src/provider/calendar/symbols.rs [hunk 3]
```diff
@@ -118,14 +123,7 @@ macro_rules! symbols {
...
     pub struct SymbolsV1<'data>(
-        #[cfg_attr(
-            feature = "serde",
-            serde(
-                borrow,
-                deserialize_with = "icu_provider::serde::borrow_de_utils::array_of_cow"
-            )
-        )]
-        pub [Cow<'data, str>; 12],
+        #[cfg_attr(feature = "serde", serde(borrow))] pub ZeroMap<'data, MonthCode, str>,
```
Comment (sffc) [inline components/datetime/src/provider/calendar/symbols.rs:126]: Thought: I hadn't considered the data size increase of this. Should we consider mitigating it by making `SymbolsV1` be an enum between the 12-month cow, like we had before, and the zeromap? Slight increase in code size but likely a more significant win in data size.

Comment (Manishearth) [reply components/datetime/src/provider/calendar/symbols.rs:126]: hmmm, we could, yes

Comment (Manishearth) [reply components/datetime/src/provider/calendar/symbols.rs:126]: Just realized that we've been omitting the 13th month for 13 month cals all along, oops

**Review by nordzilla — DISMISSED** [review]
> The regen commit is massive, avoid opening it in your browser and instead review commit by commit.

Went through each commit as suggested. Changes look reasonable to me.

**Comment by Manishearth** [comment]
Thanks! Removing dminor since I have component owner approval, but if another component owner also wishes to review I'm happy to wait for them.

File: components/datetime/src/provider/calendar/symbols.rs (hunk 1 above)
Comment (sffc) [reply components/datetime/src/provider/calendar/symbols.rs:242]: I still sort-of think that MonthCode should enforce the syntax, `/^M\d\dL?$/`, in which case a macro makes sense. But if there's no invariant, I don't think there's much of a win.

File: components/datetime/src/provider/calendar/symbols.rs [hunk 4]
```diff
@@ -117,20 +123,76 @@ macro_rules! symbols {
...
-        #[cfg_attr(
-            feature = "serde",
-            serde(
-                borrow,
-                deserialize_with = "icu_provider::serde::borrow_de_utils::array_of_cow"
-            )
-        )]
-        pub [Cow<'data, str>; 12],
-    );
+    pub enum SymbolsV1<'data> {
```
Comment (sffc) [inline components/datetime/src/provider/calendar/symbols.rs:130]: Thought: Do we want to make this an untagged enum? In terms of JSON data, it's easy to tell if this is the list version or the map version.

**Review by sffc — APPROVED** [review]

File: components/datetime/src/provider/calendar/symbols.rs (hunk 4 above)
Comment (Manishearth) [reply components/datetime/src/provider/calendar/symbols.rs:130]: We can't, that only works with self-describing formats (e.g. JSON), and will break postcard. (I tried just to be sure too). `untagged` is not a part of _serde_'s data model, it's a part of `serde_derive`'s data model, and as such serde doesn't do anything special there.

File: components/datetime/src/provider/calendar/symbols.rs (hunk 1 above)
Comment (Manishearth) [reply components/datetime/src/provider/calendar/symbols.rs:246]: I feel like perhaps that's a followup worth filing for polish, then. There may be some discussion necessary to see if we'll ever expand month code syntax with other things.

**Pull request merged**